### PR TITLE
BUG: Index.get_loc always raise InvalidIndexError on listlike

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

Shouldn't be user-facing unless user is using Index.get_loc directly since we catch this in the relevant Series/DataFrame indexing methods.  The upshot here is this will allow us to simplify `Series.__setitem__` (and maybe others)

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3599,6 +3599,10 @@ def get_loc(self, key, method=None, tolerance=None):
                 return self._engine.get_loc(casted_key)
             except KeyError as err:
                 raise KeyError(key) from err
+            except TypeError:
+                # If we have a listlike key, raise InvalidIndexError instead
+                self._check_indexing_error(key)
+                raise
```
Comment (jreback) [inline pandas/core/indexes/base.py:3605]: should be from err?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3605]: nope.  if self._check_indexing_error doesn't raise, then we're just re-raising the TypeError.  can flesh out the comment to that extent

**Comment by jbrockmendel** [comment]
mypy complaints are unrelated, just opened a PR that should address them

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3608]: fleshed out comment + green

**Pull request merged**